[LLDB] Add field member operators to DIL

cmtice opened the pull request:
Add the arrow and period operators, allowing DIL to find and access member fields.

cmtice committed:
[LLDB] Add field member operators to DIL

Add the arrow and period operators, allowing DIL to find and access
member fields.

labath commented on lldb/include/lldb/ValueObject/DILAST.h:
```diff
@@ -88,6 +89,29 @@ class IdentifierNode : public ASTNode {
+
+  ASTNode *base() const { return m_base.get(); }
+  bool IsArrow() const { return m_is_arrow; }
+  ConstString FieldName() const { return m_field_name; }
```
Let's not use ConstString here. All this does here is increase lldb memory consumption.

cmtice replied:
Done.

labath commented on lldb/include/lldb/ValueObject/DILAST.h:
```diff
@@ -88,6 +89,29 @@ class IdentifierNode : public ASTNode {
+
+  llvm::Expected<lldb::ValueObjectSP> Accept(Visitor *v) const override;
+
+  ASTNode *base() const { return m_base.get(); }
+  bool IsArrow() const { return m_is_arrow; }
```
Let's be consistent with the names here. I see we've used `GetFoo` elsewhere.

@kuilpd, I missed this in the UnaryOpNode (because it was self-consistent at least, but I think that one should also use the GetFoo scheme.

cmtice replied:
Done.

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -272,4 +272,204 @@ Interpreter::Visit(const UnaryOpNode *node) {
+lldb::ValueObjectSP
+Interpreter::EvaluateMemberOf(lldb::ValueObjectSP value,
+                              const std::vector<uint32_t> &path,
+                              bool use_synthetic, bool is_dynamic) {
```
`use_dynamic` was passed in the the parser as a `lldb::DynamicValueType`. Why did it turn into a `bool` here?

cmtice replied:
Removed this function so comment doesn't matter now.

labath reviewed (commented):
A very predictable question. Why is the implementation of this more complicated than `base->GetChildMemberWithName(name)`. I think it shouldn't be.

To avoid repeating the same points over and over, I suggest going through the discussion on the dereference/addressof patch. If you still have questions after that, let me know.

cmtice committed:
Merge remote-tracking branch 'origin/main' into DIL-member-of

cmtice committed:
Remove special code for finding field members and rely entirely on
ValueObject::GetChildMemberWithName.  Also minor code cleanups requested
by reviewer.

cmtice commented:
> A very predictable question. Why is the implementation of this more complicated than `base->GetChildMemberWithName(name)`. I think it shouldn't be.
> 
> To avoid repeating the same points over and over, I suggest going through the discussion on the dereference/addressof patch. If you still have questions after that, let me know.

GetChildMemberWithName didn't handle everything correctly. I've fixed part of that now. There's still a bug in the code that searches for fields in base classes, which I want to try to find & fix.  I've removed all the special handling code from this PR.

Please review again. Thanks!

labath commented on lldb/include/lldb/ValueObject/DILAST.h:
```diff
@@ -88,6 +89,29 @@ class IdentifierNode : public ASTNode {
+
+  ASTNode *GetBase() const { return m_base.get(); }
+  bool GetIsArrow() const { return m_is_arrow; }
+  std::string GetFieldName() const { return m_field_name; }
```
```suggestion
  llvm::StringRef GetFieldName() const { return m_field_name; }
```

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -272,4 +272,66 @@ Interpreter::Visit(const UnaryOpNode *node) {
 
+llvm::Expected<lldb::ValueObjectSP>
+Interpreter::Visit(const MemberOfNode *node) {
+  Status error;
```
Declare it where it's being used to make it clear what's its scope and contents.

cmtice replied:
Done.

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -272,4 +272,66 @@ Interpreter::Visit(const UnaryOpNode *node) {
+Interpreter::Visit(const MemberOfNode *node) {
+  Status error;
+  auto base_or_err = Evaluate(node->GetBase());
+  if (!base_or_err) {
+    return base_or_err;
+  }
```
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

cmtice replied:
Done.

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -272,4 +272,66 @@ Interpreter::Visit(const UnaryOpNode *node) {
+  CompilerType base_type = base->GetCompilerType();
+  // When using an arrow, make sure the base is a pointer or array type.
+  // When using a period, make sure the base type is NOT a pointer type.
+  if (node->GetIsArrow() && !base_type.IsPointerType() &&
+      !base_type.IsArrayType()) {
```
I think we should drop the type checks and perform the dereference unconditionally.

cmtice replied:
Why?  The current "frame var" implementation does perform these checks and return these errors...

labath replied:
Because we want to let the values (and their type systems) decide whether they are dereferencable or not. That's what we did for the implementation of the `*` operator.

I could be convinced by the "status quo" argument, but this doesn't look like an equivalent implementation to that (for one, you're adding support for dereferencing arrays), which is why I want to turn this around: if `*` does not check for types, then why should `->` (which is defined as `(*x).`) do that?

What I can imagine is doing something on the error handling path -- if dereferencing failed, we look at the value to provide a better error message about why it failed. However, even in this case, I'd try to avoid looking at the type too much and instead try to reuse the error message from the Dereference operation (changing it, if needed) so that we can be sure it makes sense regardless of how the value depends to implement dereferencing.

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -272,4 +272,66 @@ Interpreter::Visit(const UnaryOpNode *node) {
+      return llvm::make_error<DILDiagnosticError>(
+          m_expr, errMsg, node->GetLocation(), node->GetFieldName().size());
+    }
+  } else if (!node->GetIsArrow() && base_type.IsPointerType()) {
```
same here

cmtice replied:
Same question/response as before.

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -272,4 +272,66 @@ Interpreter::Visit(const UnaryOpNode *node) {
+  }
+
+  // User specified array->elem; need to get to element[0] to look for fields.
+  if (node->GetIsArrow() && base_type.IsArrayType())
+    base = base->GetChildAtIndex(0);
```
This probably shouldn't exist

cmtice replied:
Why?

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -272,4 +272,66 @@ Interpreter::Visit(const UnaryOpNode *node) {
+  lldb::ValueObjectSP field_obj =
+      base->GetChildMemberWithName(llvm::StringRef(node->GetFieldName()));
+  if (field_obj && field_obj->GetName().GetString() == node->GetFieldName()) {
+    if (field_obj->GetCompilerType().IsReferenceType()) {
+      lldb::ValueObjectSP tmp_obj = field_obj->Dereference(error);
```
Why is this necessary? GetChildMemberWithName seems to work just fine on references:
```
(lldb) script lldb.frame.FindVariable("b")
(A &) b = 0x00007fffffffd7e8: {
  X = 42
  x = 0x00007fffffffd7e8
}
(lldb) script lldb.frame.FindVariable("b").GetChildMemberWithName("X")
(int) X = 42
```

cmtice replied:
We can remove it; I'll do that.

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -272,4 +272,66 @@ Interpreter::Visit(const UnaryOpNode *node) {
+    return field_obj;
+  }
+
+  if (node->GetIsArrow() && base_type.IsPointerType())
+    base_type = base_type.GetPointeeType();
+  std::string errMsg =
+      llvm::formatv("no member named '{0}' in {1}", node->GetFieldName(),
```
What kind of error do we get from GetChildMemberWithName? Could we forward that?

cmtice replied:
We don't get an error from GetChildMemberWithName. Either it returns a ValueObjectSP, or it returns a nullptr.

labath replied:
Okay, sounds good then.

labath commented on lldb/source/ValueObject/DILParser.cpp:
```diff
@@ -111,7 +111,27 @@ ASTNodeUP DILParser::ParseUnaryExpression() {
-  return ParsePrimaryExpression();
+  return ParsePostfixExpression();
+}
+// Parse a postfix_expression.
```
```suggestion

// Parse a postfix_expression.
```

cmtice replied:
Done.

labath commented on lldb/test/API/commands/frame/var-dil/basics/MemberOf/main.cpp:
```diff
@@ -0,0 +1,59 @@
+}
+
+/*
+  EXPECT_THAT(Eval("s.x"), IsEqual("1"));
+  EXPECT_THAT(Eval("s.r"), IsEqual("2"));
+  EXPECT_THAT(Eval("s.r + 1"), IsEqual("3"));
+  EXPECT_THAT(Eval("sr.x"), IsEqual("1"));
+  EXPECT_THAT(Eval("sr.r"), IsEqual("2"));
+  EXPECT_THAT(Eval("sr.r + 1"), IsEqual("3"));
+  EXPECT_THAT(Eval("sp->x"), IsEqual("1"));
+  EXPECT_THAT(Eval("sp->r"), IsEqual("2"));
+  EXPECT_THAT(Eval("sp->r + 1"), IsEqual("3"));
+  EXPECT_THAT(Eval("sarr->x"), IsEqual("5"));
+  EXPECT_THAT(Eval("sarr->r"), IsEqual("2"));
+  EXPECT_THAT(Eval("sarr->r + 1"), IsEqual("3"));
+  EXPECT_THAT(Eval("(sarr + 1)->x"), IsEqual("1"));
+
+  EXPECT_THAT(
+      Eval("sp->4"),
+      IsError(
+          "<expr>:1:5: expected 'identifier', got: <'4' (numeric_constant)>\n"
+          "sp->4\n"
+          "    ^"));
+  EXPECT_THAT(Eval("sp->foo"), IsError("no member named 'foo' in 'Sx'"));
+  EXPECT_THAT(
+      Eval("sp->r / (void*)0"),
+      IsError("invalid operands to binary expression ('int' and 'void *')"));
+
+  EXPECT_THAT(Eval("sp.x"), IsError("member reference type 'Sx *' is a "
+                                    "pointer; did you mean to use '->'"));
+  EXPECT_THAT(
+      Eval("sarr.x"),
+      IsError(
+          "member reference base type 'Sx[2]' is not a structure or union"));
+
+  // Test for record typedefs.
+  EXPECT_THAT(Eval("sa.x"), IsEqual("3"));
+  EXPECT_THAT(Eval("sa.y"), IsEqual("'\\x04'"));
+
+*/
```
??

cmtice replied:
Whoops; notes to self re original tests in lldb-eval. Will remove that.

cmtice committed:
Minor code cleanups:
  - Formatting issues.
  - Update the evalution of MemberOf to NOT automatically deref references
  - Update tests to match.
  - Change GetFieldName to return a StringRef instead of a std::string.

cmtice committed:
- Remove code treating arrays as pointers.
- Remove code that automatically dereferenced result.
- Update member-of logic to more closely match current frame var
  implementation.

cmtice commented:
I believe I have addressed all the review comments so far. Please take another look. Thanks!

labath commented on lldb/include/lldb/ValueObject/DILAST.h:
```diff
@@ -88,6 +89,42 @@ class IdentifierNode : public ASTNode {
+  ASTNode *GetBase() const { return m_base.get(); }
+  bool GetIsArrow() const { return m_is_arrow; }
+  llvm::StringRef GetFieldName() const { return llvm::StringRef(m_field_name); }
+  bool GetCheckPtrVsMember() const { return m_check_ptr_vs_member; }
+  bool GetFragileIvar() const { return m_fragile_ivar; }
+  bool GetSynthChild() const { return m_use_synth_child; }
+  lldb::DynamicValueType GetUseDynamic() const { return m_use_dynamic; }
```
I think we discussed this (briefly) at one of the previous PRs, and I think the conclusion was that things like these should be a property of the interpreter rather than of a specific node.
I can sort of imagine a world some parts of an expression are evaluated using dynamic types and some aren't, but I don't think it's your intention to create that world. (And if it is, then these properties (at least some of them), should be other nodes as well, as e.g. `[]` also needs to know whether it should be looking at synthetic children).

cmtice replied:
Done.

labath reviewed (commented):
I think this is good strategy -- take the existing implementation of `.` and improve that later.

I have one comment about the location of the members, but otherwise, I think this looks good.

cmtice committed:
Merge remote-tracking branch 'origin/main' into DIL-member-of

cmtice committed:
Make fragile_ivar, check_ptr_vs_member and use_synthetic members
of the Interpreter.

kazutakahirata commented:
@cmtice I'm getting:

```
lldb/include/lldb/ValueObject/DILParser.h:121:8: error: private field 'm_fragile_ivar' is not used [-Werror,-Wunused-private-field]
  121 |   bool m_fragile_ivar;
      |        ^
lldb/include/lldb/ValueObject/DILParser.h:122:8: error: private field 'm_check_ptr_vs_member' is not used [-Werror,-Wunused-private-field]
  122 |   bool m_check_ptr_vs_member;
      |        ^
2 errors generated.
```

Would you mind taking a look?  Should I just remove these variables, adjust the constructor, and the callers of the constructor?  Thanks!

cmtice commented:
> @cmtice I'm getting:
> 
> ```
> lldb/include/lldb/ValueObject/DILParser.h:121:8: error: private field 'm_fragile_ivar' is not used [-Werror,-Wunused-private-field]
>   121 |   bool m_fragile_ivar;
>       |        ^
> lldb/include/lldb/ValueObject/DILParser.h:122:8: error: private field 'm_check_ptr_vs_member' is not used [-Werror,-Wunused-private-field]
>   122 |   bool m_check_ptr_vs_member;
>       |        ^
> 2 errors generated.
> ```
> 
> Would you mind taking a look? Should I just remove these variables, adjust the constructor, and the callers of the constructor? Thanks!

I'll try to get a fix for this asap.

cmtice commented:
> > @cmtice I'm getting:
> > ```
> > lldb/include/lldb/ValueObject/DILParser.h:121:8: error: private field 'm_fragile_ivar' is not used [-Werror,-Wunused-private-field]
> >   121 |   bool m_fragile_ivar;
> >       |        ^
> > lldb/include/lldb/ValueObject/DILParser.h:122:8: error: private field 'm_check_ptr_vs_member' is not used [-Werror,-Wunused-private-field]
> >   122 |   bool m_check_ptr_vs_member;
> >       |        ^
> > 2 errors generated.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Would you mind taking a look? Should I just remove these variables, adjust the constructor, and the callers of the constructor? Thanks!
> 
> I'll try to get a fix for this asap.

I think this should fix the problem: https://github.com/llvm/llvm-project/pull/141259 . Kazu, could you verify that it does please?

kazutakahirata commented:
> I think this should fix the problem: #141259 . Kazu, could you verify that it does please?

@cmtice Yes, it does.  I've LGTMed #141259.  Thank you for fixing this quickly!
